js/private/image/index.ts

```diff
@@ -0,0 +1,120 @@
+import { Archiver, create } from 'archiver'
+import { createReadStream, createWriteStream } from 'node:fs'
+import { readdir, readFile, realpath, stat } from 'node:fs/promises'
+import * as path from 'node:path'
+
+const [entriesPath, appLayerPath, nodeModulesLayerPath] = process.argv.slice(2)
+
+const app = create('tar', { gzip: true })
+app.pipe(createWriteStream(appLayerPath))
+const app_structure = new Set<string>()
+
+const node_modules = create('tar', { gzip: true })
+node_modules.pipe(createWriteStream(nodeModulesLayerPath))
+const node_modules_structure = new Set<string>()
+
+const entries: {
+    [k: string]: {
+        is_source: boolean
+        is_directory: boolean
+        dest: string
+        root?: string
+    }
+} = JSON.parse((await readFile(entriesPath)).toString())
+
+function mkdirP(p: string, output: Archiver, structure: Set<string>) {
+    const dirname = path.dirname(p).split('/')
+    let prev = '/'
+    for (const part of dirname) {
+        if (!part) {
+            continue
+        }
+        prev = path.join(prev, part)
+        if (structure.has(prev)) {
+            continue
+        }
+        structure.add(prev)
+        output = output.append(
+            null as any,
+            { name: prev, type: 'directory' } as any
+        )
+    }
+}
+
+function findKeyByValue(value: string): string {
+    for (const [key, { dest: val }] of Object.entries(entries)) {
+        if (val == value) {
+            return key
+        }
+    }
+    throw new Error(
+        `couldn't map ${value} to a path. please file a bug at https://github.com/aspect-build/rules_js/issues/new/choose`
+    )
+}
+
+async function* walk(dir: string, accumulate = '') {
+    const dirents = await readdir(dir, { withFileTypes: true })
+    for (const dirent of dirents) {
+        if (dirent.isDirectory()) {
+            yield* walk(
+                path.join(dir, dirent.name),
+                path.join(accumulate, dirent.name)
+            )
+        } else {
+            yield path.join(accumulate, dirent.name)
+        }
+    }
+}
+
+for (const key of Object.keys(entries).sort()) {
```

Comment:
this whole file needs some commenting and maybe some tests

Reply:
agreed. i'll get on that